lib/Conversion/HandshakeToHW/HandshakeToHW.cpp

```diff
@@ -1059,6 +1063,36 @@ class SelectConversionPattern
   };
 };
 
+// Rewrite handshake.instance op. Simply replace it with a hw.instance op
+class InstanceConversionPattern 
+    : public OpConversionPattern<handshake::InstanceOp> {
+public:
+  InstanceConversionPattern(ESITypeConverter &typeConverter,
+                             MLIRContext *context, 
+                             HandshakeLoweringState &ls)
+      : OpConversionPattern<handshake::InstanceOp>::OpConversionPattern(typeConverter, context),
+        ls(ls) {}
+
+  LogicalResult
+  matchAndRewrite(handshake::InstanceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    
+    // check if referenced module already exists
+    hw::HWModuleLike refModule = checkSubModuleOp(ls.parentModule, op);
+    if(!refModule)
+      assert(refModule && "handshake.instance references a module that does not exists");
```

Comment:
Is this a case that can occur when parsing input IR? If so, you should return failure in that case. An assertion should check preconditions and similar conditions that should hold in *all* cases. If an assertion is triggered, a programmer messed up. If assertions can be reached by providing certain input to an application, something is fishy.

Note that assertions will be eliminated when compiling in release mode, so this could result in undefined behaviour.

If this is indeed something that a verifier should have caught before, rewrite it to 

```
assert(!refModule && "handshake.instance references a module that does not exists");
```
without the if.

Reply:
I think this is already checked with the function InstanceOp::verifySymbolUses in the file HandshakeOps.cpp. Therefore I will rewrite it to
`assert(!refModule && "handshake.instance references a module that does not exists");`